### PR TITLE
Requesting reviews for capstone

### PR DESCRIPTION
**Pull request by Rabiann**
Hello!

This is my capstone project for the bootcamp, and I'm submitting it for review. The project is a Telegram bot that allows users to take tests, as well as create and edit them. It's a fully functional bot that I've built using Rust, Teloxide, and PostgreSQL.
Key Features:
- Users can take tests and receive immediate feedback.
- Administrators can create and edit tests through a simple interface.
- Docker is used for easy deployment and setup.

Your review is important for me to complete the [bootcamp](https://github.com/rust-lang-ua/rustcamp), and your input will be invaluable in helping me polish the final version.

Thank you in advance for taking the time to review it!

### CODE REVIEW

**Review by TheBestTvarynka — APPROVED** [review]

File: src/constructor.rs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+use teloxide::types::ReplyMarkup;
+use teloxide::{payloads::SendMessageSetters, prelude::Requester, types::Message, Bot};
+use tracing::instrument;
+use crate::database::connection::{CreateQuiz, RetreiveQuiz};
+use crate::database::quiz::{Answer, Question, Quiz};
+use crate::keyboard::{action_keyboard, yes_no_keyboard};
+use crate::state::{QuizData, QuizState};
+use crate::{HandlerResult, UserDialogue};
+
+#[instrument(level = "info", skip(connection))]
```
Comment (Citizen4our) [inline src/constructor.rs]: Do you need to logging variables Bot&dialogue all the time? it is a bit excessively

**Review by Citizen4our — APPROVED** [review]

File: src/constructor.rs (hunk 1 above)
Comment (Rabiann) [reply src/constructor.rs]: Thank you for your review. I removed bot and dialogue records from logger.

**Review by FairlyBet — APPROVED** [review]

File: Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+[dependencies]
+dotenvy = "0.15.7"
+log = "0.4.22"
+pretty_env_logger = "0.5.0"
+sqlx = { version = "0.8.2", features = ["postgres", "runtime-tokio", "uuid"] }
+teloxide = { version = "0.13.0", features = ["axum", "macros", "webhooks", "webhooks-axum"] }
+tokio = { version = "1.40.0", features = ["macros", "rt-multi-thread"] }
+tracing = "0.1.40"
+tracing-log = "0.2.0"
+tracing-subscriber = { version = "0.3.18", features = ["json", "std"] }
```
Comment (yboyi) [inline Cargo.toml]: Using both log and tracing does not seem reasonable. This can lead to code inconsistency.

File: src/main.rs [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+type HandlerResult = Result<(), Box<dyn std::error::Error + Send + Sync + 'static>>;
+
+#[tokio::main]
+async fn main() {
+    dotenv().ok();
+    // pretty_env_logger::init();
+    let rust_log = std::env::var("LOG_LEVEL").unwrap_or("error".into());
+    println!("{rust_log}");
+    tracing_subscriber::fmt().with_max_level(level_filters::LevelFilter::from_level(rust_log.parse().unwrap())).json().with_span_events(FmtSpan::ENTER).log_internal_errors(true).with_ansi(true).with_line_number(true).with_target(false).init();
+
```
Comment (yboyi) [inline src/main.rs]: Use https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html

File: src/main.rs [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+use teloxide::utils::command::BotCommands;
+use url::Url;
+
+pub mod constructor;
+pub mod database;
+pub mod keyboard;
+pub mod runner;
+pub mod state;
+pub mod editor;
+pub mod commands;
```
Comment (yboyi) [inline src/main.rs]: It's better to separate the actual logic (lib) and the executable binary (bin).
Your main.rs file should be as declarative as possible

File: src/commands.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+//         // QuizState::HandleAnswer { quiz_name, question_name, answer_name } => QuizState::HandleQuestion { quiz_name, question_name },
+//         // QuizState::EditAnswerText { quiz_name, question_name, answer_name } => QuizState::HandleAnswer { quiz_name, question_name, answer_name },
+//         // QuizState::EditCorectness { quiz_name, question_name, answer_name } => QuizState::HandleAnswer { quiz_name, question_name, answer_name },
+//         _ => unreachable!()
+//     };
+
+//     dialogue.update(to_update).await?;
+
+//     Ok(())
+// }
```
Comment (yboyi) [inline src/commands.rs]: Cluttering space

File: src/database/connection.rs [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+
+    async fn retreive_all_answers_names(&self, id_quiz: impl Into<String>, id_question: impl Into<String>) -> Result<Vec<String>, Box<dyn Error + Send + Sync>>;
+}
+
+pub(crate) trait EditAnswer {
+    async fn edit_answer_text(&self, quiz_id: impl Into<String>, question_id: impl Into<String>, answer_id: impl Into<String>, new: impl Into<String>)
+        -> GenericError;
+
+    async fn edit_corectness(&self, quiz_id: impl Into<String>, question_id: impl Into<String>, answer_id: impl Into<String>, is_correct: bool) -> GenericError;
+}
```
Comment (yboyi) [inline src/database/connection.rs]: These traits are really weird from the architecture perspective. But let it be for this PR.

**Review by yboyi — CHANGES_REQUESTED** [review]
On top of the comments I left, please, use `cargo fmt` on push, so that other devs can clearly follow the flow with the default formatting.

File: Cargo.toml (hunk 2 above)
Comment (Rabiann) [reply Cargo.toml]: Thanks, I removed log and pretty_env_logger from my deps.

File: src/main.rs (hunk 3 above)
Comment (Rabiann) [reply src/main.rs]: I used EnvFilter instead of using string.

File: src/main.rs (hunk 4 above)
Comment (Rabiann) [reply src/main.rs]: I moved my logic into separate files.

File: src/commands.rs (hunk 5 above)
Comment (Rabiann) [reply src/commands.rs]: I removed that.

File: src/database/connection.rs (hunk 6 above)
Comment (Rabiann) [reply src/database/connection.rs]: I used it as a way to establish operations, which can be done with db by handlers.

Comment (yboyi) [reply src/database/connection.rs]: > I used it as a way to establish operations, which can be done with db by handlers.

Usually, it's done by a bigger trait with all the methods. I understand that your approach can provide some additional flexibility for cases where not all connection types support the same operation set, but I wouldn't overcomplicate this part for this project.

**Review by yboyi — APPROVED** [review]
Now looks better, nice work!

**Pull request merged**